flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/functions/aggfunctions/RankLikeAggFunctionBase.java

```diff
@@ -98,30 +108,30 @@ protected Expression orderKeyEqualsExpression() {
 		return ret.orElseGet(() -> literal(true));
 	}
 
-	protected Expression generateInitLiteral(InternalType orderType) {
-		if (orderType.equals(InternalTypes.BOOLEAN)) {
+	protected Expression generateInitLiteral(LogicalType orderType) {
+		if (orderType instanceof BooleanType) {
```

Comment:
Leverage the logical type root instead of instanceof checks.

Reply:
Agree, `InternalType` is more similar to `LogicalTypeRoot`.

Reply:
Yeah, and Java can use `switch case`.

Reply:
For simple codes or codes in Java, I prefer use the logical type root.
But for complex type codes in Scala, I prefer use type to check directly. Just like:
```
t match {
  case rowType: RowType => // use rowType do some thing
  case arrayType: ArrayType => // use arrayType do some thing
}
```
I think it is better than:
```
t.getTypeRoot match {
  case LogicalTypeRoot.ROW => t.asInstanceOf[RowType] // use rowType do some thing
  case LogicalTypeRoot.ARRAY => t.asInstanceOf[ArrayType] // use rowType do some thing
}
```
What do you think?

Reply:
I would recommend to use the provided visitors instead. I also have a default visitor in my FLINK-12254 branch. FLINK-12254 will add a lot of tooling around types as well. Maybe this PR was a bit too early, but this is no problem.

Reply:
Glad to see new visitors in FLINK-12254.
I looked at the code again and found that use type to check directly is not so strong.
The below code is not bad:
```
t.getTypeRoot match {
      case BOOLEAN => s"$rowTerm.getBoolean($indexTerm)"
      case TINYINT => s"$rowTerm.getByte($indexTerm)"
      case SMALLINT => s"$rowTerm.getShort($indexTerm)"
      ......
      case ROW => s"$rowTerm.getRow($indexTerm, ${t.asInstanceOf[RowType].getFieldCount})"
}
```
So I change all LogicalType instance match to the logical type root match.